shows_app/src/app/components/feature/shows/ShowsComponent/ShowsComponent.tsx

```diff
@@ -1,24 +1,75 @@
-import React from "react";
+import React, { useEffect, useState } from "react";
+import { Flex } from "@chakra-ui/react";
 import { ShowsDetails } from "../ShowsDetails";
 import { ShowReviewSection } from "../ShowReviewSection/ShowReviewSection ";
 import { Ishows } from "@/app/typings/shows";
-import { Flex } from "@chakra-ui/react";
+import { IReview, IReviewContent } from "@/app/typings/reviews";
 
 interface IShowsComponentProps {
   show: Ishows;
 }
 
+const saveToLocalStorage = (reviews: Array<IReview>) => {
+  localStorage.setItem("reviews", JSON.stringify(reviews));
+};
+
+const loadFromLocalStorage = () => {
+  const reviewsString = localStorage.getItem("reviews");
+  if (reviewsString) {
+    return JSON.parse(reviewsString);
+  }
+  return [];
+};
+
+const calculateAverageRating = (reviews: Array<IReview>) => {
+  if (reviews.length === 0) return 0;
+  const totalRating = reviews.reduce((sum, review) => sum + review.rating, 0);
+  const showsAverageRating = totalRating / reviews.length;
+  return showsAverageRating;
+};
+
+
+
 export const ShowsComponent = ({ show }: IShowsComponentProps) => {
-  const [averageRating, setAverageRating] = React.useState<number>(0);
+  const [reviews, setReviews] = useState<Array<IReview>>([]);
+  const [averageRating, setAverageRating] = useState<number>(0);
+
+  useEffect(() => {
+    const savedReviews = loadFromLocalStorage();
+    setReviews(savedReviews);
+  }, []);
+
+  useEffect(() => {
+    saveToLocalStorage(reviews);
+    const avgRating = calculateAverageRating(reviews);
+    setAverageRating(avgRating);
```

Comment:
We are almost there. One thing that can be improved on is that you don't even need the average rating state anymore. The only time you are updating it is when the `reviews` state is changed.
A change in the `reviews` state will trigger a rerender of this component. A rerender is basically just your component (function) executing again.
So if you remove the `averageRating` useState and just do `const avgRating = calculateAverageRating(reviews)` somewhere in your component, you can pass `avgRating` to the `ShowDetails`component. That way, `avgRating` will be recalculated every time the `reviews` gets changed.

To me, this is better because you don't have to keep track of 2 states. There is only one single state (`reviews`) that is the source of truth.

Reply:
Ispravljeno u assignment-3-materials